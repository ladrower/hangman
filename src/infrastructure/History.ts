import * as H from 'history'

export interface History extends H.History {}
export class History implements H.History {}
export interface Location extends H.Location {}
export class Location implements H.Location {}
