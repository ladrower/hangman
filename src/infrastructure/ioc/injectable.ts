import 'reflect-metadata'

type IConstruct<T> = new (...args: any[]) => T

const key = Symbol('injectable')
const container = new WeakMap()

export class UndefinedDependencyException {}

export class UnresolvedDependencyException<T extends new () => any> {
  constructor(readonly dependency: T) {}
}

export const injector = {
  has<T>(type: IConstruct<T>): boolean {
    return container.has(type)
  },
  resolve<T>(constructor: IConstruct<T>, argsFrom: number): any[] | never {
    return getDependencies(constructor, argsFrom)
  },
  provide<T>(type: IConstruct<T>, value: T) {
    container.set(type, value)
    return this
  },
  invalidate<T>(type: IConstruct<T>) {
    container.delete(type)
    return this
  }
}

/**
 * @injectable class decorator
 * Forces TS to emit metadata and lazily maps instance for IoC container
 */
export function injectable<T extends IConstruct<any>>(constructor: T) {
  const C = class extends constructor {
    constructor(...args: any[]) {
      super(...args)
      const meta = Reflect.getOwnMetadata(key, C)
      if (meta.injecting && !container.has(C)) {
        container.set(C, this)
      }
    }
  }
  Reflect.defineMetadata(
    key,
    {
      injecting: false
    },
    C
  )
  Object.defineProperty(C, 'name', {
    value: constructor.name
  })
  return C
}

function getDependencies<T>(constructor: IConstruct<T>, skip: number): any[] | never {
  const types = Reflect.getMetadata('design:paramtypes', constructor) || []

  return (skip > 0 ? types.slice(skip) : types).map((type: IConstruct<any>) => {
    if (type == null) {
      throw new UndefinedDependencyException()
    }
    if (container.has(type)) {
      return container.get(type)
    }
    if (Reflect.hasOwnMetadata(key, type)) {
      const meta = Reflect.getOwnMetadata(key, type)
      meta.injecting = true
      const instance = new type(...getDependencies(type, 0))
      meta.injecting = false
      return instance
    }
    throw new UnresolvedDependencyException(type)
  })
}
