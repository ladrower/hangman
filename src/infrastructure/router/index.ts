export * from './Route'
export * from './Router'
export * from './Switch'
export { Redirect } from 'react-router'
