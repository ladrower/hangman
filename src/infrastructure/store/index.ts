export * from './observe'
export * from './store'
export * from './StoreContext'
export * from './Scope'
export * from './useStore'
