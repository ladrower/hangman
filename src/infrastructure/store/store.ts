import 'reflect-metadata'

type IConstruct<T> = new (...args: any[]) => T

let counter = 0
export function store<T extends IConstruct<any>>(constructor: T) {
  const uid = counter++
  Object.defineProperty(constructor, '::storeId', {
    value: `${constructor.name}-${uid}`
  })
  return constructor
}
