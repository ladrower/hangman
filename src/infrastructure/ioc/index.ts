import { injector } from './injectable'

export * from './injectable'

export const createNew = <T>(constructor: new (...args: any[]) => T) => {
  return new constructor(...injector.resolve(constructor, 0))
}
