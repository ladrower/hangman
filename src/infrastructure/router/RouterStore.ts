import { action } from 'mobx'
import { computed, observable } from 'mobx'
import { History, Location } from '../History'
import { store } from '@/infrastructure/store/store'

@store
export class RouterStore {
  @observable location: Location

  @computed get url() {
    const { pathname, search, hash } = this.location
    return pathname + search + hash
  }

  @action protected updateLocation = (location: Location) => (this.location = location)

  constructor(readonly history: History) {
    this.location = this.history.location
    this.history.listen(this.updateLocation)
  }

  navigate = <T extends {}>(to: string, state?: T) => this.history.push(to, state)
}
