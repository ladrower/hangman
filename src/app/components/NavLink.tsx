import React, { ReactNode, useCallback } from 'react'
import { useStore } from '@/infrastructure/store'
import { RouterStore } from '@/infrastructure/router/RouterStore'
import { Link } from '@material-ui/core'

type Props = {
  children: ReactNode,
  to: string,
}

export function NavLink({to, children}: Props) {
  const router = useStore(RouterStore)
  const handleClick = useCallback((e: React.SyntheticEvent) => {
    e.preventDefault()
    router.navigate(to)
  }, [to])

  return (
    <Link href={to} onClick={handleClick}>
      {children}
    </Link>
  )
}
