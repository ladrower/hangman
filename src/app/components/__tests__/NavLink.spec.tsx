import React from 'react'
import renderer from 'react-test-renderer'
import { NavLink } from '../NavLink'
import { Scope } from '@/infrastructure/store';
import { RouterStore } from '@/infrastructure/router/RouterStore';
import { injector } from '@/infrastructure/ioc';
import { History } from '@/infrastructure/History';
import { createMemoryHistory } from 'history';

afterEach(() => {
    injector.invalidate(History)
})

it('should render navigation link', () => {
    injector
        .provide(History, createMemoryHistory())

    const tree = renderer
        .create(
            <Scope store={RouterStore}>
                <NavLink to='/'>Home</NavLink>
            </Scope>
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})
