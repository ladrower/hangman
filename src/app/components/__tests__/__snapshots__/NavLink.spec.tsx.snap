// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render navigation link 1`] = `
<a
  className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
  href="/"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
>
  Home
</a>
`;
